cip/CIP2015-06-24-call-procedures.adoc

```diff
@@ -0,0 +1,250 @@
+For such a case, this CIP suggests that arguments with default values are written as `<name> = <value> :: <type`.
+
+==== Procedure implementation
+
+In an actual implementation, procedures may be written in different programming languages and provided with APIs for accessing the underlying graph database or graph processing system as appropriate.
+The details of this are outside of the scope of this proposal.
+
+=== Procedure calls
+
+Procedures are called for each incoming record produced by preceding clauses and each time produce arbitrary many result rows. If a statement starts with a procedure call, this call is executed once, i.e. procedure calls behave similiar to `UNWIND` regarding how they change the result cardinality.
+
+We propose new syntax that introduces a new clause for calling procedures.
+This clause for calling procedures starts with the `CALL` keyword.
+Next follows the actual call (i.e. procedure name and arguments in signature order).
+Finally (and optionally) the clause may end with the `YIELD` keyword followed by any chosen result field projections.
+
+A procedure call constitutes either an entire Cypher statement (i.e. is a "standalone call"), or is called as part of a larger query (i.e. is an "in-query call"). In standalone calls, arguments may be either passed implicitly or explicitly. These different calling modes are detailed in the following subsections.
+
+==== In-query calls
+
+Calling a procedure as part of a larger query follows three basic rules:
+
+- All arguments are always passed explicitly in the order given by the signature.
+- Result fields are projected and appended as new variable bindings explicitly in the order given by the procedure call.
+- New variable bindings projected from result fields are not allowed to shadow existing variable bindings that are already in scope.
+
+These rules ensure that looking at a procedure call provides enough information to ascertain it's impact on the variable scope in it's part of the query without having to be aware of any other already bound variables.
+
+As a basic example, consider a call to the stored procedure `myProc(name``::``STRING?``,``id``::``INTEGER?)``::`` (last``::``STRING?)`:
+
+.Calling stored procedure inside a larger query
+[source, cypher]
+----
+MATCH (n:Person)-[r:IN]->(g:Group)
+CALL myProc(n.name, g.id*1000+r.id) YIELD last AS lastLogin
+RETURN *
+----
+
+This calls `myProc` for each record produced by the preceding `MATCH` clause with a `name` argument obtained by evaluating `n.name` and an `id` argument obtained by evaluating `g.id*1000+r.id`.
+Each call may yield multiple procedure result records with a single result field `last`.
+For each of them, a new statement result record is created that contains the original variables already in scope (i.e. `n`, `r`, `g`) as well as the projected result field `last` renamed as `lastLogin`.
+
+`YIELD -` is used to indicate that no result fields should be projected.
```

Comment:
Please add wording here to the effect that the call will still affect the cardinality of the query